paddle/fluid/operators/inplace_abn_op.cc

```diff
@@ -0,0 +1,205 @@
+                      "Scale input should be of float type");
+    PADDLE_ENFORCE_EQ(bn_param_type, ctx.Input<Tensor>("Bias")->type(),
+                      "Bias input should be of float type");
+    PADDLE_ENFORCE_EQ(bn_param_type, ctx.Input<Tensor>("Mean")->type(),
+                      "Mean input should be of float type");
+    PADDLE_ENFORCE_EQ(bn_param_type, ctx.Input<Tensor>("Variance")->type(),
+                      "Variance input should be of float type");
+
+    framework::LibraryType library = framework::LibraryType::kPlain;
+    framework::DataLayout layout = framework::DataLayout::kAnyLayout;
+
+    return framework::OpKernelType(input_data_type, ctx.GetPlace(), layout,
+                                   library);
+  }
+};
+
+class InplaceABNGradOp : public paddle::operators::BatchNormGradOp {
+ public:
+  using paddle::operators::BatchNormGradOp::BatchNormGradOp;
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    const auto* var = ctx.InputVar(framework::GradVarName("Y"));
+    auto input_data_type = ctx.Input<Tensor>("X")->type();
+    if (var == nullptr) {
+      PADDLE_THROW("can't find Y@GRAD");
+    }
+    const Tensor* t = nullptr;
+    if (var->IsType<Tensor>()) {
+      t = &var->Get<Tensor>();
+    } else if (var->IsType<LoDTensor>()) {
+      t = &var->Get<LoDTensor>();
+    }
+    if (t == nullptr) {
+      PADDLE_THROW("can't find Y@GRAD");
+    }
+    framework::LibraryType library = framework::LibraryType::kPlain;
+    framework::DataLayout layout = framework::DataLayout::kAnyLayout;
+
+    return framework::OpKernelType(input_data_type, ctx.GetPlace(), layout,
+                                   library);
+  }
+};
+
+class InplaceABNOpMaker : public paddle::operators::BatchNormOpMaker {
+ public:
+  void Make() override {
+    BatchNormOpMaker::Make();
+    AddAttr<std::string>(
+        "activation",
+        "(enum string, default identity, can be identity|elu|leakyrelu) "
+        "The activation type used for output candidate {h}_t.")
+        .SetDefault("");
+    AddAttr<float>("alpha",
+                   "(float, default 1.0) Only used in inplace-abn kernel,"
+                   "the activation type(identity|elu|leakyrelu) would be fused "
+                   "with batch_norm, "
+                   "this is the alpha value for elu|leakyrelu.")
+        .SetDefault(0.1f);
+    AddAttr<bool>("use_sync_bn",
+                  "(bool, default false) Whether use synchronize batch "
+                  "normalization.")
+        .SetDefault(false);
+  }
+};
+
+template <typename T>
+class InplaceABNOpGradMaker : public framework::SingleGradOpMaker<T> {
+ public:
+  using framework::SingleGradOpMaker<T>::SingleGradOpMaker;
+
+ protected:
+  std::unique_ptr<T> Apply() const override {
+    auto* op = new T();
+    op->SetType(this->ForwardOpType() + "_grad");
+    op->SetInput("X", this->Input("X"));
```

Comment:
Why add `X`? Seems that it makes the forward part `inplace_abn` cannot perform inplace.